CMake build should work in parallel

We need to ensure the build can work in parallel e.g. `make -j2` and `ninja`.

Last time I tested `ninja` on Linux, it didn't work (though I've heard the Windows build so far is working with Ninja / VS+CMake).

Here's an example of one failure:

```
creating config.mak... done
[  4%] Performing build step for 'musl_includes'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[  4%] No install step for 'musl_includes'
[  4%] Completed 'musl_includes'
[  4%] Built target musl_includes
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/3rdparty/mbedtls/CMakeLists.txt
+++ b/3rdparty/mbedtls/CMakeLists.txt
@@ -31,7 +31,11 @@ ExternalProject_Add(mbedtls-wrap
     -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
 
   # This is a no-op to override the default.
-  INSTALL_COMMAND "")
+  INSTALL_COMMAND ""
+  BYPRODUCTS
+    <BINARY_DIR>/library/libmbedtls.a
+    <BINARY_DIR>/library/libmbedx509.a
+    <BINARY_DIR>/library/libmbedcrypto.a)
 
 ExternalProject_Get_property(mbedtls-wrap SOURCE_DIR)
 ExternalProject_Get_property(mbedtls-wrap BINARY_DIR)

--- a/tools/oeedger8r/CMakeLists.txt
+++ b/tools/oeedger8r/CMakeLists.txt
@@ -7,21 +7,34 @@ else()
     set(BINARY ${OE_BINDIR}/oeedger8r)
 endif()
 
+# NOTE: The custom commands below first copy the input files to the
+# build directory and then invoke the OCaml tools because those tools
+# do not emit to the current working directory, they always emit to
+# the location of the input file.
+
 # Generate Lexer.
+find_program(OCAMLLEX ocamllex)
+if (NOT OCAMLLEX)
+  message(FATAL_ERROR "ocamllex not found! Please install 'ocaml-native-compilers'")
+endif ()
+
 add_custom_command(
-    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/Lexer.ml
-    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/intel/Lexer.mll ${CMAKE_CURRENT_BINARY_DIR}
-    COMMAND ocamllex Lexer.mll
-    DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/intel/Lexer.mll 
-)
+  OUTPUT Lexer.ml
+  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/intel/Lexer.mll ${CMAKE_CURRENT_BINARY_DIR}
+  COMMAND ocamllex Lexer.mll
+  DEPENDS intel/Lexer.mll)
 
 # Generate Parser.
+find_program(OCAMLYACC ocamlyacc)
+if (NOT OCAMLYACC)
+  message(FATAL_ERROR "ocamlycac not found! Please install 'ocaml-native-compilers'")
+endif ()
+
 add_custom_command(
-    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/Parser.ml ${CMAKE_CURRENT_BINARY_DIR}/Parser.mli
-    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/intel/Parser.mly ${CMAKE_CURRENT_BINARY_DIR}
-    COMMAND ocamlyacc Parser.mly
-    DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/intel/Parser.mly 
-)
+  OUTPUT Parser.ml Parser.mli
+  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/intel/Parser.mly ${CMAKE_CURRENT_BINARY_DIR}
+  COMMAND ocamlyacc Parser.mly
+  DEPENDS intel/Parser.mly)
 
 # Compile
 add_custom_command(
@@ -38,13 +51,14 @@ add_custom_command(
     COMMAND ocamlopt -c -I . -o Emitter.cmx ${CMAKE_CURRENT_SOURCE_DIR}/Emitter.ml
     COMMAND ocamlopt -c -I . -o main.cmx ${CMAKE_CURRENT_SOURCE_DIR}/main.ml
     COMMAND ocamlopt -c -I . -o Emitter.cmx ${CMAKE_CURRENT_SOURCE_DIR}/Emitter.ml
-    COMMAND ocamlopt str.cmxa unix.cmxa Ast.cmx Util.cmx SimpleStack.cmx Plugin.cmx 
+    COMMAND ocamlopt str.cmxa unix.cmxa Ast.cmx Util.cmx SimpleStack.cmx Plugin.cmx
                      Preprocessor.cmx Parser.cmx Lexer.cmx CodeGen.cmx Emitter.cmx main.cmx
                      -o ${BINARY}
 
     # Add dependency to generated Lexer and Parser, and all of the sources.
-    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/Lexer.ml 
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/Lexer.ml
             ${CMAKE_CURRENT_BINARY_DIR}/Parser.ml
+            ${CMAKE_CURRENT_BINARY_DIR}/Parser.mli
             Emitter.ml
             main.ml
             intel/Ast.ml
@@ -56,9 +70,7 @@ add_custom_command(
             intel/Util.ml
 )
 
-add_custom_target(oeedger8r ALL
-    DEPENDS ${BINARY}
-)
+add_custom_target(oeedger8r ALL DEPENDS ${BINARY})
 
 # install rule
 install (PROGRAMS ${BINARY} DESTINATION ${CMAKE_INSTALL_BINDIR})

